src/Faker/Calculator/Luhn.php

```diff
@@ -55,4 +58,41 @@ public static function isValid($number)
     {
         return self::checksum($number) === 0;
     }
+
+    /**
+     * Generate a Luhn compliant number.
+     *
+     * @param string $prefix
+     * @param int $length
+     * @return string
+     */
+    public static function generateLuhnNumber($prefix, $length)
+    {
+        $length = intval($length);
+        $prefix = trim($prefix);
+
+        if (!preg_match('/^\d*$/', $prefix)) {
+            throw new InvalidArgumentException('prefix should be all digits.');
+        }
+
+        if ($length <= 0) {
+            throw new InvalidArgumentException('length should be greater than zero.');
+        }
+
+        if (strlen($prefix) > $length) {
+            throw new InvalidArgumentException('prefix should be longer than the length.');
+        }
+
+        if (strlen($prefix) == $length) {
+            if (static::isValid($prefix)) {
+                return $prefix;
+            } else {
+                throw new InvalidArgumentException('prefix length equals length but prefix is not a Luhn number.');
+            }
+        }
+
+        $pattern = $prefix . str_repeat('#', $length - strlen($prefix) - 1);
+        $partialValue = BaseProvider::numerify($pattern);
```

Comment:
you should pass the partialValue as parameter to this method, to separate the responsibilities: Luhn calculation (this class), random generation (Company.php)